src/libraries/mysterium-client/monitoring.js

```
@@ -1,55 +1,95 @@
+  _subscribersUp: Array<UpCallback> = []
+  _subscribersDown: Array<DownCallback> = []
+
+  constructor (tequilapi: TequilapiClient) {
     this.api = tequilapi
-    this.clientIsRunning = false
-    this.apiTimeout = null
-    this.ipcTimeout = null
   }
 
   start () {
-    this._healthCheck()
+    this._healthCheckLoop()
   }
 
   stop () {
-    if (this.apiTimeout) {
-      clearTimeout(this.apiTimeout)
-    }
-    if (this.ipcTimeout) {
-      clearTimeout(this.ipcTimeout)
+    if (this._timer) {
+      clearTimeout(this._timer)
     }
   }
 
-  isRunning () {
-    return this.clientIsRunning
+  subscribeStatus (callback: StatusCallback) {
+    this._subscribersStatus.push(callback)
+  }
+
+  subscribeUp (callback: UpCallback) {
+    this._subscribersUp.push(callback)
   }
 
-  onProcessReady (callback) {
-    const interval = setInterval(() => {
-      if (this.clientIsRunning) {
-        clearInterval(interval)
-        callback()
-      }
-    }, 100)
+  subscribeDown (callback: DownCallback) {
+    this._subscribersDown.push(callback)
   }
 
-  async _healthCheck () {
+  async _healthCheckLoop (): Promise<boolean> {
+    let isRunning
     try {
       await this.api.healthCheck(healthCheckTimeout)
-      this.clientIsRunning = true
+      isRunning = true
     } catch (e) {
-      this.clientIsRunning = false
+      isRunning = false
+    }
+
+    this._notifySubscribers(isRunning)
+
+    this._timer = setTimeout(() => this._healthCheckLoop(), healthCheckInterval)
+  }
+
+  _notifySubscribers (isRunning: boolean) {
+    this._notifySubscribersStatus(isRunning)
+
+    if (this._lastIsRunning === isRunning) {
+      return
+    }
+    if (isRunning) {
+      this._notifySubscribersUp()
+    } else {
+      this._notifySubscribersDown()
     }
+    this._lastIsRunning = isRunning
+  }
+
+  _notifySubscribersStatus (isRunning: boolean) {
+    this._subscribersStatus.forEach((callback: StatusCallback) => {
+      try { callback(isRunning) } catch (e) {}
```

We shouldn't ignore errors of listeners - we should `logger.info` and `bugReporter.captureInfoException` them.
Same applies for `_notifySubscribersUp` and `_notifySubscribersDown`

---

Not my scope. Fixing in #184